File: java-study/java-syntax/src/main/java/algorithm/SortAndRanking.java
package algorithm;

import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.ArrayList;
import java.util.Collections;

public class SortAndRanking {
    public static void main(String[] args) {
        Student[] students = {
                new Student("Maria", 5),
                new Student("John", 9),
                new Student("Paul", 7),
                new Student("Ada", 5),
        };

        System.out.println("Students without sort: " + Arrays.toString(students));
        
        Arrays.sort(students);
        System.out.println("Students sorted by score: " + Arrays.toString(students));

        Comparator<Student> studentComparator = Comparator
            .comparingInt(Student::getScore)
            .thenComparing(Student::getName);

        Arrays.sort(students, studentComparator);
        System.out.println("Students sorted by score and name: " + Arrays.toString(students));

        for (int i = 0; i < students.length; i++) {
            students[i].setRank(i + 1);
        }

        Arrays.sort(students, (a, b) -> Integer.compare(a.getRank(), b.getRank()));
        System.out.println("Students sorted by rank: " + Arrays.toString(students));

        List<Student> studentList = Arrays.asList(students);
        System.out.println("Students array list: " + studentList);

        Collections.sort(studentList, (a, b) -> Integer.compare(b.getRank(), a.getRank()));
        System.out.println("Students array list sorted by rank descending: " + studentList);

        // ArrayList.sort
        // Collections.sort
        // Arrays.sort
    }
}

class Student implements Comparable<Student> {
    private String name;
    private Integer score;
    private Integer rank;

    public Student(String name, Integer score) {
        this.name = name;
        this.score = score;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Integer getScore() {
        return score;
    }

    public void setScore(Integer score) {
        this.score = score;
    }

    public Integer getRank() {
        return rank;
    }

    public void setRank(Integer rank) {
        this.rank = rank;
    }

    public int compareTo(Student other) {
        return Integer.compare(this.score, other.score);
    }

    public String toString() {
        return "[" + this.name +
            ", " + this.score +
            ", " + this.rank + "]";
    }
}

File: kotlin-study/kotlin-syntax/src/main/kotlin/algorithm/SortAndRanking.kt

fun main() {
  val students = arrayOf(
    Student("Maria", 5),
    Student("John", 9),
    Student("Paul", 7),
    Student("Ada", 5),
  )

  println("Students unsorted: ${students.joinToString()}")

  students.sort()

  println("Students soted by score: ${students.joinToString()}")

  // students.sortWith(compareByDescending { it.score })
  students.sortWith(compareBy({it.score}, {it.name}))

  println("Students by score and name: ${students.joinToString()}")

  students.sortBy { it.name }

  println("Students by name: ${students.joinToString()}")

  //for ( (index, item) in students.withIndex()) { item.rank = index + 1 }
  students.forEachIndexed { index, item -> item.rank = index + 1 }
  students.sortBy { it.rank }

  println("Students by rank: ${students.joinToString()}")

}

data class Student(
  val name: String,
  val score: Int,
  var rank: Int? = 0,
) : Comparable<Student> {
  override fun compareTo(other: Student) : Int = this.score.compareTo(other.score)
  override fun toString() : String = "name: $name, score: $score, rank: $rank"
}
